src/Roslyn.Diagnostics.Analyzers/Core/AbstractApplyTraitToClass.cs

```diff
@@ -0,0 +1,111 @@
+            if (attributeType.Type is not { Name: "TraitAttribute", ContainingNamespace: { Name: "Xunit", ContainingNamespace: { IsGlobalNamespace: true } } })
+                return;
+
+            var location = attribute.GetLocation();
+            context.RegisterRefactoring(
+                CodeAction.Create(
+                    RoslynDiagnosticsAnalyzersResources.ApplyTraitToContainingType,
+                    cancellationToken => ApplyTraitToClassAsync(context.Document, location.SourceSpan, cancellationToken),
+                    nameof(AbstractApplyTraitToClass<TAttributeSyntax>)));
+        }
+
+        private async Task<Document> ApplyTraitToClassAsync(Document document, TextSpan sourceSpan, CancellationToken cancellationToken)
+        {
+            var semanticModel = (await document.GetSemanticModelAsync(cancellationToken).ConfigureAwait(false))!;
+
+            var syntaxTree = await document.GetSyntaxTreeAsync(cancellationToken).ConfigureAwait(false);
+            var syntaxRoot = await syntaxTree!.GetRootAsync(cancellationToken).ConfigureAwait(false);
+            var reportedNode = syntaxRoot.FindNode(sourceSpan, getInnermostNodeForTie: true);
+            var typeDeclaration = GetTypeDeclarationForNode(reportedNode);
+            if (typeDeclaration is null)
+                return document;
+
+            var syntaxGenerator = SyntaxGenerator.GetGenerator(document);
+            if (syntaxGenerator.TryGetContainingDeclaration(reportedNode, DeclarationKind.Attribute) is not { } attribute)
+            {
+                return document;
+            }
+
+            if (syntaxGenerator.TryGetContainingDeclaration(attribute, DeclarationKind.Method) is not { } method)
+            {
+                // The attribute is not applied to a method
+                return document;
+            }
+
+            var expectedAttributeData = semanticModel.GetDeclaredSymbol(method, cancellationToken).GetAttributes()
+                .Single(attributeData => attributeData.ApplicationSyntaxReference is not null && attribute.Span.Contains(attributeData.ApplicationSyntaxReference.Span));
+
+            var newTypeDeclaration = typeDeclaration.ReplaceNodes(
+                syntaxGenerator.GetMembers(typeDeclaration),
+                (originalNode, replacementNode) =>
+                {
+                    foreach (var attribute in syntaxGenerator.GetAttributes(originalNode))
+                    {
+                        var attributeType = semanticModel.GetTypeInfo(attribute, cancellationToken);
+                        if (attributeType.Type is null)
+                            attributeType = semanticModel.GetTypeInfo(attribute.ChildNodes().First(), cancellationToken);
+
+                        if (attributeType.Type is not { Name: "TraitAttribute", ContainingNamespace: { Name: "Xunit", ContainingNamespace: { IsGlobalNamespace: true } } })
```

Comment:
consider extracting helper for this check.

Reply:
@mavasani do we have a helper for this already in the repository?

Reply:
⏱ Fixing this by using `GetOrCreateTypeByMetadataName`